### PR TITLE
 Add `ring_buffer` module with generic `Fixed` and `Bounded` ring buffer types

### LINKED ISSUE #56: Ring Buffer type
It would be great to make our own ring buffer (basically, something like `VecDeque`) that uses a `<F: Frame> &mut [F]` in its constructor. (It's impossible to do this with `VecDeque`. I would consider this a 1.0 milestone feature for being able to pitch this library.

### PR DESCRIPTION
**Pull request by mitchmindtree**
`Fixed` is a ring buffer of fixed length, meaning each time an element
is `push`ed to the back, an element must also be popped from the front.

`Bounded` is a ring buffer with a fixed **max** length and allows for
pushing and popping to be performed separately.

Both types can be created from any type that implements the provided
`ring_buffer::Slice` and `SliceMut` traits. These traits are implemented
for all types that can provide a `slice` that can be used for pushing
and popping elements. Implementations are provided for fixed-size arrays
(1...128, 256, 512, 1024, 2048, 4096, 8192), `Vec`, `Box<[T]>` and
`&mut [T]`.

All public methods have documentation explaining behaviour and most have
examples that demonstrate use.

Some further tests have been added to the test module.

cc @andrewcsmith would be great to get your thoughts on this!

Closes #56.

### CODE REVIEW

**Comment by andrewcsmith** [comment]
I have two initial thoughts, neither of which are blockers for this new feature, but I would want to take both into consideration before switching chunks of my own code to this implementation.

1. *Performance and alignment of values.* [bounded_spsc_queue does some crazy stuff](https://github.com/polyfractal/bounded-spsc-queue/blob/master/src/lib.rs#L38) to make sure there's padding and to benchmark as-fast-as-possible access to the various values. I find (worst-case) performance to be very important in this case. 

2. *Producer / Consumer model.* There needs to be some way of splitting the producer and consumer and transferring each to a separate thread. That said, maybe this isn't necessary. I could see still using `bounded_spsc_queue` for the channel use-case, and then using `sample::ring_buffer` for internal buffers on things like buses and interpolation.

That said, it looks here like `Bounded` and `Fixed` are lighter weight than the multithreaded ring buffer that I've been using for most purposes. So, basically, I'm interested to see whether this changes things to use this as a thread-internal ring buffer.

Anyway, long story short, I don't see any red flags. I'm happy with the scope of it and will give it a go in my next project or refactor.

**Comment by mitchmindtree** [comment]
> Performance and alignment of values.

I agree, it would be great to start building a benchmark suite!

> Producer / Consumer model.

I should have mentioned that `Fixed` and `Bounded` are intended only for single-threaded use-cases. Perhaps we can look into the possibility of adding `spsc` or even `mpsc` queues in the future, though I'd imagine it would only be worth it (over using types from other crates) if there was some optimisation we could perform for `Copy` types, as most of the `sample` queue/ringbuffer intended use-cases will be for samples/frames I'd imagine.

Thanks for taking a look!
